README.md
# BrainFunc
Do `stack run filename.b` to run a bf file.

(Made as a quick fun project in Spring 2022.)


ChangeLog.md
# Changelog for BrainFunc

## Unreleased changes


app/Main.hs
module Main where

import BFEvalBase
import BFRep
import BFTokenEval
import System.Environment
import System.IO

{-
1. Load the BF file into memory
2. Translate the BF file into an IR
3. Simplify the IR (maybe during step 2?)
4. Output a "compiled" file
-}

main :: IO ()
main = do
    hSetBuffering stdout NoBuffering
    args <- getArgs
    if (args == []) 
        then error "No input file given"
    else do
        let filepath = head args
        program <- readFile filepath
        -- UNCOMMENT THESE LINES TO DO NAIVE EVAL
        -- let mach = initMachine (toBFChars program)
        -- outState <- naiveEval mach
        -- UNCOMMENT THESE LINES TO DO FANCY EVAL
        let mach = initTokMachine (makeCollapsedTokens program)
        outMach <- eval mach
        putStrLn "\n"
        return ()


src/BFRep.hs
module BFRep where
    -- TODO: set what functions this exports

import Data.Maybe
import Debug.Trace

data BFChar = BPlus | BMinus | BLeft | BRight | BIn | BOut | BLBrack | BRBrack | BEnd deriving (Eq)

instance Show BFChar where
    show BPlus   = "+"
    show BMinus  = "-"
    show BLeft   = "<"
    show BRight  = ">"
    show BIn     = ","
    show BOut    = "."
    show BLBrack = "["
    show BRBrack = "]"
    show BEnd    = "!"

toBFChars :: String -> [BFChar]
toBFChars = mapMaybe toBFChar

toBFChar :: Char -> Maybe BFChar
toBFChar c = case c of
    '+' -> Just BPlus
    '-' -> Just BMinus
    '<' -> Just BLeft
    '>' -> Just BRight
    ',' -> Just BIn
    '.' -> Just BOut
    '[' -> Just BLBrack
    ']' -> Just BRBrack
    _   -> Nothing

-- Now let's do a different approach which should condense the programs being run

-- oh no, BFTs
data BFToken = BtokAdd Int | BtokMov Int | BtokIn | BtokOut | BtokWhile [BFToken] | BtokEnd deriving (Eq)

instance Show BFToken where
    show (BtokAdd i) = "add " ++ show i
    show (BtokMov i) = "mov " ++ show i
    show (BtokIn)    = "input"
    show (BtokOut)   = "print"
    show (BtokWhile toks) = "while {" ++ show toks ++ "}"
    show (BtokEnd)   = "end"

safeHead :: [a] -> Maybe a
safeHead [] = Nothing
safeHead xs = Just (head xs)

safeTail :: [a] -> [a]
safeTail [] = []
safeTail xs = tail xs

-- fun idea about bftokens: maybe we can change these back to bchars
-- program synthesis as a future direction???

-- toBFTokensH prog = (remainingProg, token)
-- NOTE: seenToks is reversed! 
toBFToken :: String -> (String, Maybe BFToken)
toBFToken ""       = ("", Just BtokEnd)
toBFToken ('[':cs) =  (fst b, Just $ BtokWhile (snd b))
    where b = toBFTokensH cs
toBFToken ('+':cs) = (cs, Just $ BtokAdd 1)
toBFToken ('-':cs) = (cs, Just $ BtokAdd (-1))
toBFToken ('>':cs) = (cs, Just $ BtokMov 1)
toBFToken ('<':cs) = (cs, Just $ BtokMov (-1))
toBFToken ('.':cs) = (cs, Just BtokOut)
toBFToken (',':cs) = (cs, Just BtokIn)
-- toBFToken (']':cs) = (cs, Just BtokEnd)
toBFToken (_:cs)   = (cs, Nothing)

toBFTokensH :: String -> (String, [BFToken])
toBFTokensH ""       = ("", [])
toBFTokensH (']':cs) = (cs, [])
toBFTokensH str      = (rcs, alltoks)
    where 
     rcs = fst restParsed
     alltoks = case thisToken of
         BtokEnd -> snd restParsed
         tok     -> tok:(snd restParsed)
     restParsed = toBFTokensH $ fst getThisToken
     thisToken = case snd getThisToken of
         Just j -> j
         Nothing -> BtokEnd
     getThisToken = toBFToken str

toBFTokens :: String -> [BFToken]
toBFTokens s = snd $ toBFTokensH s

collapseTokensList :: [BFToken] -> [BFToken]
-- TODO: catch unmatched brackets here!
collapseTokensList toks = case toks of 
    (BtokAdd n) : (BtokAdd m) : rest -> collapseTokensList $ (BtokAdd $ n + m):rest
    (BtokMov n) : (BtokMov m) : rest -> collapseTokensList $ (BtokMov $ n + m):rest
    tok : rest -> tok:(collapseTokensList rest)
    [] -> []

makeCollapsedTokens :: String -> [BFToken]
makeCollapsedTokens = collapseTokensList . toBFTokens

-- >>> let prog = "+++-+<+"
-- >>> makeCollapsedTokens prog
-- [add 3,mov -1,add 1]
--




src/BFEval.hs
module BFEval where

import BFRep
import Control.Monad.State
import Data.Array
import Data.Char
import Data.Map (Map, lookup, findWithDefault, insert, empty)
import Text.Read

{-
TODO: Rewrite this description

Evaluation requires
    1. a snapshot of the tape
    2. the program
    3. PC and tape head location
    4. the current instruction

Furthermore, we need a stack of bracket locations if where
want the bracket handling to not be absolutely awful.

TODO: Update! We're using lookup tables instead of a stack
-}

type Stack = [Int]

data BFMachine = BFMachine {
    prog :: Array Int BFChar,  -- the program is immutable :)
    cyc  :: Int,
    tape :: Map Int Int,
    pc :: Int,
    wh :: Int,
    brax :: Map Int Int
}

instance Show BFMachine where
    show (BFMachine p y t c w b) =
        ("CYCLE " ++ show y) ++ "\n" ++
        ("Tape: " ++ show t) ++ "\n" ++
        ("PC: " ++ show c) ++ "\n" ++
        ("WH: " ++ show w) ++ "\n"
        

-- kinda bad implementation of matching brackets
makeBracketsMap :: [BFChar] -> Maybe (Map Int Int)
makeBracketsMap cs = makeBracketsMapH cs 0 [] (Just empty)

-- makeBracketsMapH progList ind lbrackStack mapping = newMapping
makeBracketsMapH :: [BFChar] -> Int -> [Int] -> Maybe (Map Int Int) -> Maybe (Map Int Int)
makeBracketsMapH (BRBrack:cs) _ [] _ = Nothing
makeBracketsMapH [] _ (b:bs) _       = Nothing
makeBracketsMapH [] _ _ (Just m)            = Just m
makeBracketsMapH (c:cs) i bx (Just m)   = case c of
    BLBrack -> makeBracketsMapH cs (i+1) (i:bx) (Just m)
    BRBrack -> case bx of 
        b:bs -> makeBracketsMapH cs (i+1) bs $ Just (insert b i (insert i b m))
        -- note that if bx were nil, we'd have caught it already!
    _       -> makeBracketsMapH cs (i+1) (bx) (Just m)

indPair :: [a] -> [(Int, a)]
indPair ls = zip [0..(length ls - 1)] ls

initMachine :: [BFChar] -> BFMachine
initMachine pl = (BFMachine p 0 empty 0 0 $! b) where
    p = array (0, length pl) (indPair pl ++ [(length pl, BEnd)])
    b = case makeBracketsMap pl of
        Just m -> m
        Nothing -> error "Compile error: Imbalanced brackets" 

getCurrCmd :: StateT BFMachine IO BFChar
getCurrCmd = state (\ (BFMachine p y t c w b) -> ((p!c), BFMachine p y t c w b))

tapeWrite :: Int -> Int -> StateT BFMachine IO ()
tapeWrite loc n = state (\ (BFMachine p y t c w b) -> ((), BFMachine p y (insert loc n t) c w b))

tapeRead :: Int -> StateT BFMachine IO Int
tapeRead loc = state (\ (BFMachine p y t c w b) -> (findWithDefault 0 loc t, BFMachine p y t c w b))

writePC :: Int -> StateT BFMachine IO ()
writePC n = state (\ (BFMachine p y t c w b) -> ((), BFMachine p y t n w b))

getPC :: StateT BFMachine IO Int
getPC = state (\ (BFMachine p y t c w b) -> (c, BFMachine p y t c w b))

writeCyc :: Int -> StateT BFMachine IO ()
writeCyc n = state (\ (BFMachine p y t c w b) -> ((), BFMachine p n t c w b))

getCyc :: StateT BFMachine IO Int
getCyc = state (\ (BFMachine p y t c w b) -> (y, BFMachine p y t c w b))

advancePC :: StateT BFMachine IO ()
advancePC = do
    pc <- getPC
    writePC (pc+1)

writeWH :: Int -> StateT BFMachine IO ()
writeWH n = state (\ (BFMachine p y t c w b) -> ((), BFMachine p y t c n b))

getWH :: StateT BFMachine IO Int
getWH = state (\ (BFMachine p y t c w b) -> (w, BFMachine p y t c w b))

incCyc :: StateT BFMachine IO ()
incCyc = do
    cyc <- getCyc
    writeCyc (cyc+1)

getBrax :: StateT BFMachine IO (Map Int Int)
getBrax = state (\ (BFMachine p y t c w b) -> (b, BFMachine p y t c w b))

getBracketMatch :: Int -> StateT BFMachine IO Int
getBracketMatch loc = do
    b <- getBrax
    return $ case Data.Map.lookup loc b of 
        Just loc' -> loc'
        Nothing   -> error $ "Error finding bracket match " ++ (show loc)

-- Naive evaluation: just work on the basic BFChars.

evalWithoutAdvance :: BFChar -> StateT BFMachine IO ()
evalWithoutAdvance BPlus = do
    writeLoc <- getWH
    val <- tapeRead writeLoc
    tapeWrite writeLoc ((val+1) `mod` 256)
evalWithoutAdvance BMinus = do
    writeLoc <- getWH
    val <- tapeRead writeLoc
    tapeWrite writeLoc ((val-1) `mod` 256)
evalWithoutAdvance BLeft = do
    wh <- getWH
    writeWH (wh-1)
evalWithoutAdvance BRight = do
    wh <- getWH
    writeWH (wh+1)
evalWithoutAdvance BLBrack = do
    pc <- getPC
    wh <- getWH
    val <- tapeRead wh
    targ <- getBracketMatch pc
    when (val == 0) (writePC targ)
evalWithoutAdvance BRBrack = do
    pc <- getPC
    targ <- getBracketMatch pc
    writePC (targ - 1)    
evalWithoutAdvance BIn = do
    wh <- getWH
    liftIO (putStr "Input (0-255): ")
    inp <- liftIO getLine
    -- error (show inp)
    case (readMaybe inp) of 
        Just n  -> tapeWrite wh n
        Nothing -> error "Invalid input"
evalWithoutAdvance BOut = do
    writeLoc <- getWH
    val <- tapeRead writeLoc
    liftIO (putChar (chr val))

naiveEvalR :: StateT BFMachine IO ()
naiveEvalR = do
    -- for debugging: print state each inst
    -- mach <- Control.Monad.State.get
    -- Control.Monad.State.lift $ putStrLn (show mach)
    -- [end debug stuff]
    cmd <- getCurrCmd
    if (cmd == BEnd)
        then return ()
    else do 
        evalWithoutAdvance cmd
        advancePC
        incCyc
        naiveEvalR

naiveEval :: BFMachine -> IO BFMachine
naiveEval mach = execStateT naiveEvalR mach

src/BFTokenEval.hs
{-# LANGUAGE BangPatterns, StrictData #-}
module BFTokenEval where
    -- TODO: set what functions this exports

import BFRep
import Control.Monad.State
import Data.Char
import Data.Map.Strict as M
import Text.Read

data BFTokMachine = BFTokMachine {
    -- TODO: It seems like a list should be fine for the program?
    prog :: [BFToken],
    -- TODO: do we want "cycles"?
    tape :: Map Int Int,
    -- no need for a PC: the PC will never go backward!
    wh :: Int
    -- no need to know about brackets!
}

instance Show BFTokMachine where
    show (BFTokMachine p t w) =
        ("Tape: " ++ show t) ++ "\n" ++
        ("WH: " ++ show w) ++ "\n"

initTokMachine :: [BFToken] -> BFTokMachine
initTokMachine pl = BFTokMachine pl empty 0

readCurCell :: BFTokMachine -> Int
readCurCell mach = M.findWithDefault 0 (wh mach) (tape mach)

-- NOTE: this advances the "PC"
writeCurCell :: BFTokMachine -> Int -> BFTokMachine
writeCurCell mach n = BFTokMachine (tail $ prog mach) (insert (wh mach) (n `mod` 256) (tape mach)) (wh mach)

-- TODO: what are the consequences of not using state here?
evalTok :: BFTokMachine -> IO BFTokMachine
evalTok !mach = do
    let t = tape mach
    let w = wh mach
    case (prog mach) of
        (BtokAdd n):rest -> do
            let curVal = readCurCell mach
            return $ writeCurCell mach (curVal+n)
        (BtokMov n):rest -> 
            return $ BFTokMachine rest t (w+n)
        (BtokIn):rest -> do
            putStr "Input (0-255): "
            inp <- getLine
            case readMaybe inp of
                Just n -> return $ writeCurCell mach n
                Nothing -> error "Invalid input"
        (BtokOut):rest -> do
            putChar (chr (readCurCell mach))
            return $ BFTokMachine rest t w  -- TODO OH NO
        (BtokWhile block):rest -> do
            case readCurCell mach of
                0 -> return $ BFTokMachine rest t w
                _ -> do
                    machAfterIter <- eval $ BFTokMachine block t w
                    let newTape = tape machAfterIter
                    let newWh = wh machAfterIter
                    return $ BFTokMachine ((BtokWhile block):rest) newTape w                    

eval :: BFTokMachine -> IO BFTokMachine
eval !mach = do
    case (prog mach) of
        [] -> return mach
        p -> do
            afterMach <- evalTok mach
            eval afterMach

src/BFEvalBase.hs
module BFEvalBase where
    -- TODO: set what functions this exports

import BFRep
import Control.Monad.State
import Data.Array
import Data.Char
import Data.Map.Strict as M (Map, lookup, findWithDefault, insert, empty)
import Text.Read

{-
TODO: Rewrite this description

Evaluation requires
    1. a snapshot of the tape
    2. the program
    3. PC and tape head location
    4. the current instruction

Furthermore, we need a stack of bracket locations if where
want the bracket handling to not be absolutely awful.

TODO: Update! We're using lookup tables instead of a stack
-}

data BFMachine = BFMachine {
    prog :: Array Int BFChar,  -- the program is immutable :)
    cyc  :: Int,
    tape :: Map Int Int,
    pc :: Int,
    wh :: Int,
    brax :: Map Int Int
}

instance Show BFMachine where
    show (BFMachine p y t c w b) =
        ("CYCLE " ++ show y) ++ "\n" ++
        ("Tape: " ++ show t) ++ "\n" ++
        ("PC: " ++ show c) ++ "\n" ++
        ("WH: " ++ show w) ++ "\n"
        

-- kinda bad implementation of matching brackets
makeBracketsMap :: [BFChar] -> Maybe (Map Int Int)
makeBracketsMap cs = makeBracketsMapH cs 0 [] (Just empty)

-- makeBracketsMapH progList ind lbrackStack mapping = newMapping
makeBracketsMapH :: [BFChar] -> Int -> [Int] -> Maybe (Map Int Int) -> Maybe (Map Int Int)
makeBracketsMapH (BRBrack:cs) _ [] _ = Nothing
makeBracketsMapH [] _ (b:bs) _       = Nothing
makeBracketsMapH [] _ _ (Just m)            = Just m
makeBracketsMapH (c:cs) i bx (Just m)   = case c of
    BLBrack -> makeBracketsMapH cs (i+1) (i:bx) (Just m)
    BRBrack -> case bx of 
        b:bs -> makeBracketsMapH cs (i+1) bs $ Just (insert b i (insert i b m))
        -- note that if bx were nil, we'd have caught it already!
    _       -> makeBracketsMapH cs (i+1) (bx) (Just m)

indPair :: [a] -> [(Int, a)]
indPair ls = zip [0..(length ls - 1)] ls

initMachine :: [BFChar] -> BFMachine
initMachine pl = (BFMachine p 0 empty 0 0 $! b) where
    p = array (0, length pl) (indPair pl ++ [(length pl, BEnd)])
    b = case makeBracketsMap pl of
        Just m -> m
        Nothing -> error "Compile error: Imbalanced brackets" 

getCurrCmd :: StateT BFMachine IO BFChar
getCurrCmd = state (\ (BFMachine p y t c w b) -> ((p!c), BFMachine p y t c w b))

tapeWrite :: Int -> Int -> StateT BFMachine IO ()
tapeWrite loc n = state (\ (BFMachine p y t c w b) -> ((), BFMachine p y (insert loc n t) c w b))

tapeRead :: Int -> StateT BFMachine IO Int
tapeRead loc = state (\ (BFMachine p y t c w b) -> (findWithDefault 0 loc t, BFMachine p y t c w b))

writePC :: Int -> StateT BFMachine IO ()
writePC n = state (\ (BFMachine p y t c w b) -> ((), BFMachine p y t n w b))

getPC :: StateT BFMachine IO Int
getPC = state (\ (BFMachine p y t c w b) -> (c, BFMachine p y t c w b))

writeCyc :: Int -> StateT BFMachine IO ()
writeCyc n = state (\ (BFMachine p y t c w b) -> ((), BFMachine p n t c w b))

getCyc :: StateT BFMachine IO Int
getCyc = state (\ (BFMachine p y t c w b) -> (y, BFMachine p y t c w b))

advancePC :: StateT BFMachine IO ()
advancePC = do
    pc <- getPC
    writePC (pc+1)

writeWH :: Int -> StateT BFMachine IO ()
writeWH n = state (\ (BFMachine p y t c w b) -> ((), BFMachine p y t c n b))

getWH :: StateT BFMachine IO Int
getWH = state (\ (BFMachine p y t c w b) -> (w, BFMachine p y t c w b))

incCyc :: StateT BFMachine IO ()
incCyc = do
    cyc <- getCyc
    writeCyc (cyc+1)

getBrax :: StateT BFMachine IO (Map Int Int)
getBrax = state (\ (BFMachine p y t c w b) -> (b, BFMachine p y t c w b))

getBracketMatch :: Int -> StateT BFMachine IO Int
getBracketMatch loc = do
    b <- getBrax
    return $ case M.lookup loc b of 
        Just loc' -> loc'
        Nothing   -> error $ "Error finding bracket match " ++ (show loc)

-- Naive evaluation: just work on the basic BFChars.

evalWithoutAdvance :: BFChar -> StateT BFMachine IO ()
evalWithoutAdvance BPlus = do
    writeLoc <- getWH
    val <- tapeRead writeLoc
    tapeWrite writeLoc ((val+1) `mod` 256)
evalWithoutAdvance BMinus = do
    writeLoc <- getWH
    val <- tapeRead writeLoc
    tapeWrite writeLoc ((val-1) `mod` 256)
evalWithoutAdvance BLeft = do
    wh <- getWH
    writeWH (wh-1)
evalWithoutAdvance BRight = do
    wh <- getWH
    writeWH (wh+1)
evalWithoutAdvance BLBrack = do
    pc <- getPC
    wh <- getWH
    val <- tapeRead wh
    targ <- getBracketMatch pc
    when (val == 0) (writePC targ)
evalWithoutAdvance BRBrack = do
    pc <- getPC
    targ <- getBracketMatch pc
    writePC (targ - 1)    
evalWithoutAdvance BIn = do
    wh <- getWH
    liftIO (putStr "Input (0-255): ")
    inp <- liftIO getLine
    -- error (show inp)
    case (readMaybe inp) of 
        Just n  -> tapeWrite wh n
        Nothing -> error "Invalid input"
evalWithoutAdvance BOut = do
    writeLoc <- getWH
    val <- tapeRead writeLoc
    liftIO (putChar (chr val))

naiveEvalR :: StateT BFMachine IO ()
naiveEvalR = do
    -- for debugging: print state each inst
    -- mach <- Control.Monad.State.get
    -- Control.Monad.State.lift $ putStrLn (show mach)
    -- [end debug stuff]
    cmd <- getCurrCmd
    if (cmd == BEnd)
        then return ()
    else do 
        evalWithoutAdvance cmd
        advancePC
        incCyc
        naiveEvalR

naiveEval :: BFMachine -> IO BFMachine
naiveEval mach = execStateT naiveEvalR mach